### DISCUSSION TITLE
doing auth with fastapi + idom

### DISCUSSION THREAD

**Discussion by acivitillo** [idom-team/idom#594] [Question]
Completely experimental, never seen this before, but it works. The idea is to move auth to the FastAPI API layer and store the `access_token` in the idom state.

Is it crazy though? I mean it's relatively simple to reason about in idom, but I wonder if this makes sense. I guess the user would have to login every time the websocket connection is lost.

Better ideas? It would probably be convenient for idom to store some state in a client cookie.

```python
from fastapi import FastAPI
import requests
from idom import component, hooks, html, run, vdom, web
from idom.server.fastapi import PerClientStateServer


@component
def Authenticate(username: str = "", password: str = ""):
    user, set_state = hooks.use_state({"name": username, "password": password})

    if username != "":
        print(username, password)
        data = requests.post(
            f"http://127.0.0.1:8000/token",
            data={"username": username, "password": password},
            headers={
                "accept": "application/json",
                "Content-Type": "application/x-www-form-urlencoded",
            },
        )
        user["access_token"] = data.json()["access_token"]
        set_state(user)
        return Result(user)
    else:
        return html.div(
            html.form(
                html.input({"name": "username"}),
                html.input({"name": "password"}),
                html.button("Login"),
            ),
            Result(user),
        )


@component
def Result(user):
    try:
        data = requests.get(
            "http://127.0.0.1:8000/users/me",
            headers={"Authorization": f"Bearer {user['access_token']}"},
        )
        return html.p(data, user["name"])
    except KeyError:
        return html.p("")


run(Authenticate, port=3000)
```

**Comment by rmorshea** [comment]
There should definitely be a better way of adding authentication in IDOM. @Archmonger is using a similar strategy in a project based on [django-idom](https://github.com/idom-team/django-idom). This works in Django IDOM for two reasons:

1. We're passing info about the [websocket connection](https://github.com/idom-team/django-idom/blob/main/src/django_idom/websocket_consumer.py#L21-L26) as an argument to the root component
2. This still requires some sort of static login page where you save auth state in the client (correct me if I'm wrong @Archmonger).

We can solve **1** in IDOM core with the [upcoming](https://github.com/idom-team/idom/pull/585) `use_context` hook:

```python
from idom.server.fastapi import RequestContext
from idom import component, use_context

@component
def AuthenticatedView():
    request = use_context(RequestContext)
    # use some sort of auth token on the request
```

For **2** though, I haven't really had time to figure out a good way of saving client-side state. It's certainly possible, the question is what the API should be:

- [Custom components](https://github.com/idom-team/django-idom/blob/main/src/django_idom/websocket_consumer.py#L21-L26) for each auth solution?
- A generic way of saving client-side state?
- Built-in auth solutions for each server implementation?

Not having thought about this much, I'd lean towards the custom component solution, at least initially, since that wouldn't require any big changes.

**Reply by Archmonger** [reply]
My login is currently fully HTTP and skips IDOM.

In terms of logout...

On the Django side, I'm handling a lot of stuff by creating event driven invisible iframes. Loading the iframe triggers server side logout, and then a client side page refresh (on the parent frame).

It's janky right now but we'll figure out a better solution with time.

**Reply by acivitillo** [reply]
@Archmonger how does your HTTP login work? How will IDOM components know the user is logged?

**Reply by rmorshea** [reply]
Presently, IDOM does not expose the request associated with the underlying websocket connection. Ultimately I intend to do this with [contexts](https://github.com/idom-team/idom/pull/585). A usage might look like:

```python
from idom import component, use_context
from idom.server.fastapi import RequestContext

@component
def MyComponent():
    request = use_context(RequestContext)
    ...
```

Unfortunately I've been bogged down with work and more pressing bug fixes so I haven't managed to get contexts across the finish line.

**Reply by Archmonger** [reply]
Within Django, the HTTP session is shared with the websocket session. So authenticating on a traditional HTTP page also authenticates a websocket indirectly.

The second part to that story is Django IDOM propagates the request context via positional args. That allows me to do things such as accessing the authenticated user's object.

**Reply by Archmonger** [reply]
Also @rmorshea to clarify your "statement 2" on this post, I'm doing login entirely through traditional Django HTTP.
I don't have to do this though, since Django websockets support [login/logout methods](https://channels.readthedocs.io/en/stable/topics/authentication.html).

I choose this route since there's added efficiency in serving the login page staticly (mostly caching related).

**Comment by acivitillo** [comment]
thanks for this, sounds like what I need. Question, what will the RequestContext object contain? 


>

**Reply by Archmonger** [reply]
It's likely going to vary from framework to framework. At least within Django-IDOM it currently contains a [`scope`](https://channels.readthedocs.io/en/stable/topics/consumers.html#scope) and some basic websocket methods such as `disconnect`

**Reply by rmorshea** [reply]
> It's likely going to vary from framework to framework.

Correct. For FastAPI we will probably give direct access to the `starlette.websockets.WebScoket` object which itself contains the ASGI scope.

**Reply by acivitillo** [reply]
so @rmorshea let me understand the flow

1. the user logs via a normal fastapi page
2. on successful login fastapi stores a token into the user cookie
3. the Websocket scope now has all the info stored in the cookie
4. so our scope populates our context state, now we can use that state in the app for auth and other things

correct?

**Reply by rmorshea** [reply]
Correct.

There ought to be a more integrated approach to authentication in the long-term, but I'm not especially knowledgeable when it comes to security, and so would be nervous about implementing anything myself without guidance from someone with more expertise.

**Comment by acivitillo** [comment]
I have noticed `use_context` was released, is it possible to use it with `fastapi` or `sanic` already to fetch the websocket context?

**Reply by Archmonger** [reply]
Not yet. See the following issue:
- https://github.com/idom-team/idom/issues/669

**Reply by acivitillo** [reply]
Great. So once we have this `RequestContext` object, theoretically we could also use `use_state` and save the `scope` in state variables, right? What would be the main difference between `use_state` and `use_context` then?

**Reply by rmorshea** [reply]
You won't actually need to store the request in `use_state` - the `use_context` hook will give you global access to the request object wherever it's needed in your application. Here's a section of React's docs which discuses the reasons why you might want to create and use contexts: https://beta.reactjs.org/learn/passing-data-deeply-with-context

**Reply by rmorshea** [reply]
For example, you would be able to write:

```python
from idom.server.fastapi import RequestContext

@component
def App():
    return ...  # uses DeeplyNestedChild 

@component
def DeeplyNestedChild():
    """A child the lives somehwere deep within your App"""
    request = use_context(RequestContext)
```

**Reply by acivitillo** [reply]
I see. So the main difference is convenience and clean code with `use_context`. Whenever I need the `scope` I can just call `request = use_context(RequestContext)` and I will get it, without having to pass arguments down the components tree as I would have to do with `use_state`.

And this is clearly visible in your example because `def DeeplyNestedChild()` doesn't take any arguments.

**Reply by rmorshea** [reply]
Exactly!

**Reply by sylvoslee** [reply]
Will this be resolved before you go on vacation, or what is the interim solution

**Reply by rmorshea** [reply]
I do plan to resolve this before then. I'm planning on implementing the [RequestContext](https://github.com/idom-team/idom/issues/669) after [this PR](https://github.com/idom-team/idom/pull/703), but if having a RequestContext is blocking for anyone, I can refocus on.

**Reply by acivitillo** [reply]
@sylvoslee on which server are you running IDOM? I am using FastAPI. After @rmorshea merges [his PR](https://github.com/idom-team/idom/pull/703) I can write a tutorial on auth for IDOM if it helps. My current thinking is that it's better to do auth on the server framework and pass the auth details to the **embedded** IDOM app/s. Note 1 site could have n IDOM apps all requiring the same auth details.

**Reply by sylvoslee** [reply]
I am using FastAPI，How did you achieve it?

**Reply by acivitillo** [reply]
You do auth in FastAPI and then [embed ](https://idom-docs.herokuapp.com/docs/guides/getting-started/running-idom.html?highlight=embed#embed-in-an-existing-webpage) IDOM app inside a FastAPI page. This will work after [RequestContext](https://github.com/idom-team/idom/issues/669) is implemented, which should be happening in the next 3-4 weeks.

**Reply by rmorshea** [reply]
I've actually just decided to implement the RequestContext as part of the PR I'm currently working on - it happens to be expedient since I can remove some tests for features the context makes obsolete. So it should land a bit sooner.

**Reply by rmorshea** [reply]
Once done, something like the following should work:

```python
from fastapi import FastAPI

from idom import component, server

app = FastAPI()

@component
def MyComponent():
    conn = server.fastapi.use_connection()
    conn.user
    conn.auth
    ...
    
add_auth_to_app_somehow(app)
server.fastapi.configure(app, MyComponent)
```

Where `conn` is a `fastapi.websockets.WebSocket` object (inherits from `fastapi.requests.HTTPConnection`).

**Reply by Archmonger** [reply]
@rmorshea 

I wouldn't recommend `connection` as they keyword here. A lot of web frameworks reserve the word "connection" for database tasks. Especially when not prefixed (ex. `HttpConnection`). So `use_connection` can feel a bit awkward in some circumstances. 

Here's some alternatives that don't encroach onto some common namespaces

- `use_websocket`
   - Most frameworks contain connection scope/context within a `Websocket` object, so it makes sense to follow that naming pattern.
- `use_ws_connection`
   - This might not be wise if we ever decide to utilize HTTP instead of WS in the future. But that's currently not on IDOM's radar.
- `use_session`
   - This makes sense verbally, but might feel weird with [other upcoming hooks](https://github.com/idom-team/django-idom/issues/32).
- `use_scope`
   - Not sure how I feel about this one.

**Reply by acivitillo** [reply]
maybe session? `server.fastapi.use_session()`

**Reply by Archmonger** [reply]
Websocket and session are the two I'm leaning towards.

Slightly biased towards websocket due to preexisting conventions.

**Reply by rmorshea** [reply]
Not all implementations will necessarily give access to the websocket (e.g. Flask and Tornado expose "request" objects). While I happen to expose the "websocket" object in most cases, the minimum would be for the implementation to expose the ASGI/WSGI scope. This scope, from the ASGI docs, "is a dict containing details about the specific connection" - hence `use_connection`. My feeling is that if there are other "connections" being used in the same area of code, users should appropriately name their variables. `use_session` doesn't really avoid this need for appropriate variable names since "session" is also generic enough to have other meanings (e.g. database sessions in `sqlalchemy`).

**Reply by Archmonger** [reply]
In retrospect, I think access to the scope is all that is needed. I don't think a user would ever need to manually use WS methods such as disconnect.

So perhaps `use_scope`?

**Reply by rmorshea** [reply]
To that end, maybe the solution here is to require that all implementations have a `use_cope` hook that literally returns the WSGI/ASGI scope dict. An implementation may provide more hooks (e.g. `use_websocket` or `use_request`), but those are not guaranteed to be exist for all implementations.

**Reply by Archmonger** [reply]
Agreed, but if we're expanding out past one function I'd propose the syntax for this to be 

```python
from idom.server.flask import Scope

scope = use_context(Scope)
```

**Reply by rmorshea** [reply]
Good idea.

**Reply by rmorshea** [reply]
So, turns out it's expedient to only require the existence of a `use_scope` hook since some implementations should just do:

```python
def use_websocket() -> WebSocket:
    """Get the current WebSocket object"""
    websocket = use_context(WebSocketContext)
    if websocket is None:
        raise RuntimeError("No websocket. Are you running a Starllette server?")
    return websocket


def use_scope() -> Scope:
    """Get the current ASGI scope dictionary"""
    return use_websocket().scope
```

**Reply by rmorshea** [reply]
The work on this is nearly complete - just updating docs to reflect the changes.

**Comment by acivitillo** [comment]
I was wondering if we could handle auth in a FastApi application and somehow pass the logged user information down to IDOM? 

https://idom-docs.herokuapp.com/docs/guides/getting-started/running-idom.html?highlight=embed#embed-in-an-existing-webpage

**Comment by sylvoslee** [comment]
On fastapi, I have used middleware to get the request header https://github.com/tomwojcik/starlette-context

**Comment by rmorshea** [comment]
@acivitillo have you gotten a chance to try out 0.38.0-a1 and if so, would that release be sufficient for you to proceed with development while I'm on vacation? I'm thinking that for a 0.38.0 release I'd like to complete https://github.com/idom-team/idom/pull/721, https://github.com/idom-team/idom/issues/720, and maybe https://github.com/idom-team/idom/issues/653.

**Reply by acivitillo** [reply]
we are testing it with @liberty-rising. The use case we are working on is user logging to a FastApi page via GitHub login and passing all the GitHub user info we receive from the login to embedded IDOM thru `use_scope`. I think we'll need another 2-3 working days to give you an answer, but it looks promising as we were able to get basic auth working through `use_scope` very quickly.
